New crashes when checking Cassandra after upgrade to 3.36

Hey, while the previously filed crashes have been resolved (thanks!), there appeared new ones:

```
    [javac] error: SourceChecker.typeProcess: unexpected Throwable (IndexOutOfBoundsException) while processing /home/jlewandowski/dev/cassandra/c18239-static-analysis/src/java/org/apache/cassandra/io/util/FileUtils.java; message: -1
    [javac]   ; The Checker Framework crashed.  Please report the crash.
    [javac]   Compilation unit: /home/jlewandowski/dev/cassandra/c18239-static-analysis/src/java/org/apache/cassandra/io/util/FileUtils.java
    [javac]   Last visited tree at line 71 column 1:
    [javac]   public final class FileUtils
    [javac]   Exception: java.lang.IndexOutOfBoundsException: -1; java.lang.IndexOutOfBoundsException: -1
    [javac]   	at jdk.compiler/com.sun.tools.javac.util.List.get(List.java:482)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.convertCallArguments(CFGTranslationPhaseOne.java:1274)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitMethodInvocation(CFGTranslationPhaseOne.java:1478)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitMethodInvocation(CFGTranslationPhaseOne.java:201)
    [javac]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1666)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.scan(CFGTranslationPhaseOne.java:556)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.translateAssignment(CFGTranslationPhaseOne.java:1665)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitVariable(CFGTranslationPhaseOne.java:4019)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitVariable(CFGTranslationPhaseOne.java:201)
    [javac]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCVariableDecl.accept(JCTree.java:990)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.scan(CFGTranslationPhaseOne.java:556)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitBlock(CFGTranslationPhaseOne.java:2222)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitBlock(CFGTranslationPhaseOne.java:201)
    [javac]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1048)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.scan(CFGTranslationPhaseOne.java:556)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitTry(CFGTranslationPhaseOne.java:3557)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitTry(CFGTranslationPhaseOne.java:201)
    [javac]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCTry.accept(JCTree.java:1344)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.scan(CFGTranslationPhaseOne.java:556)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitBlock(CFGTranslationPhaseOne.java:2222)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.visitBlock(CFGTranslationPhaseOne.java:201)
    [javac]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1048)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.scan(CFGTranslationPhaseOne.java:556)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.process(CFGTranslationPhaseOne.java:455)
    [javac]   	at org.checkerframework.dataflow.cfg.builder.CFGTranslationPhaseOne.process(CFGTranslationPhaseOne.java:507)
    [javac]   	at org.checkerframework.framework.flow.CFCFGBuilder.build(CFCFGBuilder.java:73)
    [javac]   	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1529)
    [javac]   	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:1444)
    [javac]   	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1933)
    [javac]   	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:423)
    [javac]   	at org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:536)
    [javac]   	at org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:184)
    [javac]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
    [javac]   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
    [javac]   	at org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:86)
    [javac]   	at org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:1043)
    [javac]   	at org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:558)
    [javac]   	at org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:551)
    [javac]   	at org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:188)
    [javac]   	at jdk.compiler/com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:828)
    [javac]   	at jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:132)
    [javac]   	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1414)
    [javac]   	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1371)
    [javac]   	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:973)
    [javac]   	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:311)
    [javac]   	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:170)
    [javac]   	at jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:57)
    [javac]   	at jdk.compiler/com.sun.tools.javac.Main.main(Main.java:43)
```

It can be reproduced on this commit: https://github.com/jacek-lewandowski/cassandra/commit/7b3c4ce8e5d6da41791492903ec367b8857cd57d with the following command:

```
ant cf-only -Dcf.check.only=org/apache/cassandra/io/util/FileUtils.java
```


We're also getting crashes for two more files:

```
    [javac] error: SourceChecker.typeProcess: unexpected Throwable (IndexOutOfBoundsException) while processing /home/jlewandowski/dev/cassandra/c18239-static-analysis/src/java/org/apache/cassandra/io/sstable/format/big/BigFormat.java; message: -1
    [javac]   ; The Checker Framework crashed.  Please report the crash.
    [javac]   Compilation unit: /home/jlewandowski/dev/cassandra/c18239-static-analysis/src/java/org/apache/cassandra/io/sstable/format/big/BigFormat.java
    [javac]   Last visited tree at line 71 column 1:
    [javac]   public class BigFormat extends AbstractSSTableFormat<BigTableReader, BigTableWriter>
    [javac]   Exception: java.lang.IndexOutOfBoundsException: -1; java.lang.IndexOutOfBoundsException: -1
    [javac]   	at jdk.compiler/com.sun.tools.javac.util.List.get(List.java:482)
    [javac]   	at org.checkerframework.framework.type.TypeFromTypeTreeVisitor.getTypeVariableFromDeclaration(TypeFromTypeTreeVisitor.java:275)
    [javac]   	at org.checkerframework.framework.type.TypeFromTypeTreeVisitor.visitIdentifier(TypeFromTypeTreeVisitor.java:310)
    [javac]   	at org.checkerframework.framework.type.TypeFromTypeTreeVisitor.visitIdentifier(TypeFromTypeTreeVisitor.java:50)
    [javac]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCIdent.accept(JCTree.java:2272)
    [javac]   	at jdk.compiler/com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:80)
    [javac]   	at org.checkerframework.framework.type.TypeFromTree.fromTypeTree(TypeFromTree.java:77)
    [javac]   	at org.checkerframework.framework.type.TypeFromMemberVisitor.visitVariable(TypeFromMemberVisitor.java:49)
    [javac]   	at org.checkerframework.framework.type.TypeFromMemberVisitor.visitVariable(TypeFromMemberVisitor.java:29)
    [javac]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCVariableDecl.accept(JCTree.java:990)
    [javac]   	at jdk.compiler/com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:80)
    [javac]   	at org.checkerframework.framework.type.TypeFromTree.fromMember(TypeFromTree.java:63)
    [javac]   	at org.checkerframework.framework.type.AnnotatedTypeFactory.fromMember(AnnotatedTypeFactory.java:1587)
    [javac]   	at org.checkerframework.framework.type.AnnotatedTypeFactory.fromElement(AnnotatedTypeFactory.java:1516)
    [javac]   	at org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:1304)
    [javac]   	at org.checkerframework.checker.resourceleak.MustCallConsistencyAnalyzer$Obligation.getMustCallValue(MustCallConsistencyAnalyzer.java:341)
    [javac]   	at org.checkerframework.checker.resourceleak.MustCallConsistencyAnalyzer$Obligation.getMustCallMethods(MustCallConsistencyAnalyzer.java:308)
    [javac]   	at org.checkerframework.checker.resourceleak.MustCallConsistencyAnalyzer.checkMustCall(MustCallConsistencyAnalyzer.java:2075)
    [javac]   	at org.checkerframework.checker.resourceleak.MustCallConsistencyAnalyzer.propagateObligationsToSuccessorBlocks(MustCallConsistencyAnalyzer.java:1943)
    [javac]   	at org.checkerframework.checker.resourceleak.MustCallConsistencyAnalyzer.analyze(MustCallConsistencyAnalyzer.java:547)
    [javac]   	at org.checkerframework.checker.resourceleak.ResourceLeakAnnotatedTypeFactory.postAnalyze(ResourceLeakAnnotatedTypeFactory.java:138)
    [javac]   	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1606)
    [javac]   	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:1444)
    [javac]   	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1933)
    [javac]   	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:423)
    [javac]   	at org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:536)
    [javac]   	at org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:184)
    [javac]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
    [javac]   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
    [javac]   	at org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:86)
    [javac]   	at org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:1043)
    [javac]   	at org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:558)
    [javac]   	at org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:188)
    [javac]   	at jdk.compiler/com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:828)
    [javac]   	at jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:132)
    [javac]   	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1414)
    [javac]   	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1371)
    [javac]   	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:973)
    [javac]   	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:311)
    [javac]   	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:170)
    [javac]   	at jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:57)
    [javac]   	at jdk.compiler/com.sun.tools.javac.Main.main(Main.java:43)
    [javac] /home/jlewandowski/dev/cassandra/c18239-static-analysis/src/java/org/apache/cassandra/io/sstable/format/big/BigTableWriter.java:430: error: [mustcall:bound] incompatible bounds in type parameter
    [javac]         private <@MustCallAlias T> T ensuringInBuildInternalContext(T value)
    [javac]                  ^
    [javac]   type: T/*DECL*/ [ extends @MustCall Object super @PolyMustCall Void]
    [javac]   upper bound: @MustCall Object
    [javac]   lower bound: @PolyMustCall NullType
    [javac] error: SourceChecker.typeProcess: unexpected Throwable (IndexOutOfBoundsException) while processing /home/jlewandowski/dev/cassandra/c18239-static-analysis/src/java/org/apache/cassandra/io/sstable/format/bti/BtiFormat.java; message: -1
    [javac]   ; The Checker Framework crashed.  Please report the crash.
    [javac]   Compilation unit: /home/jlewandowski/dev/cassandra/c18239-static-analysis/src/java/org/apache/cassandra/io/sstable/format/bti/BtiFormat.java
    [javac]   Last visited tree at line 60 column 1:
    [javac]   public class BtiFormat extends AbstractSSTableFormat<BtiTableReader, BtiTableWriter>
    [javac]   Exception: java.lang.IndexOutOfBoundsException: -1; java.lang.IndexOutOfBoundsException: -1
    [javac]   	at jdk.compiler/com.sun.tools.javac.util.List.get(List.java:482)
    [javac]   	at org.checkerframework.framework.type.TypeFromTypeTreeVisitor.getTypeVariableFromDeclaration(TypeFromTypeTreeVisitor.java:275)
    [javac]   	at org.checkerframework.framework.type.TypeFromTypeTreeVisitor.visitIdentifier(TypeFromTypeTreeVisitor.java:310)
    [javac]   	at org.checkerframework.framework.type.TypeFromTypeTreeVisitor.visitIdentifier(TypeFromTypeTreeVisitor.java:50)
    [javac]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCIdent.accept(JCTree.java:2272)
    [javac]   	at jdk.compiler/com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:80)
    [javac]   	at org.checkerframework.framework.type.TypeFromTree.fromTypeTree(TypeFromTree.java:77)
    [javac]   	at org.checkerframework.framework.type.TypeFromMemberVisitor.visitVariable(TypeFromMemberVisitor.java:49)
    [javac]   	at org.checkerframework.framework.type.TypeFromMemberVisitor.visitVariable(TypeFromMemberVisitor.java:29)
    [javac]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCVariableDecl.accept(JCTree.java:990)
    [javac]   	at jdk.compiler/com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:80)
    [javac]   	at org.checkerframework.framework.type.TypeFromTree.fromMember(TypeFromTree.java:63)
    [javac]   	at org.checkerframework.framework.type.AnnotatedTypeFactory.fromMember(AnnotatedTypeFactory.java:1587)
    [javac]   	at org.checkerframework.framework.type.AnnotatedTypeFactory.fromElement(AnnotatedTypeFactory.java:1516)
    [javac]   	at org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:1304)
    [javac]   	at org.checkerframework.checker.resourceleak.MustCallConsistencyAnalyzer$Obligation.getMustCallValue(MustCallConsistencyAnalyzer.java:341)
    [javac]   	at org.checkerframework.checker.resourceleak.MustCallConsistencyAnalyzer$Obligation.getMustCallMethods(MustCallConsistencyAnalyzer.java:308)
    [javac]   	at org.checkerframework.checker.resourceleak.MustCallConsistencyAnalyzer.checkMustCall(MustCallConsistencyAnalyzer.java:2075)
    [javac]   	at org.checkerframework.checker.resourceleak.MustCallConsistencyAnalyzer.propagateObligationsToSuccessorBlocks(MustCallConsistencyAnalyzer.java:1943)
    [javac]   	at org.checkerframework.checker.resourceleak.MustCallConsistencyAnalyzer.analyze(MustCallConsistencyAnalyzer.java:547)
    [javac]   	at org.checkerframework.checker.resourceleak.ResourceLeakAnnotatedTypeFactory.postAnalyze(ResourceLeakAnnotatedTypeFactory.java:138)
    [javac]   	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1606)
    [javac]   	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:1444)
    [javac]   	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1933)
    [javac]   	at org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:423)
    [javac]   	at org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:536)
    [javac]   	at org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:184)
    [javac]   	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
    [javac]   	at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
    [javac]   	at org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:86)
    [javac]   	at org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:1043)
    [javac]   	at org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:558)
    [javac]   	at org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:188)
    [javac]   	at jdk.compiler/com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:828)
    [javac]   	at jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:132)
    [javac]   	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1414)
    [javac]   	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1371)
    [javac]   	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:973)
    [javac]   	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:311)
    [javac]   	at jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:170)
    [javac]   	at jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:57)
    [javac]   	at jdk.compiler/com.sun.tools.javac.Main.main(Main.java:43)
```

They cannot be reproduced in isolation - that is, you need to run `ant cf` to check all the sources to encounter the above exception; when validating only those files, the check passes.

--- a/framework/src/main/java/org/checkerframework/framework/type/TypeFromTypeTreeVisitor.java
+++ b/framework/src/main/java/org/checkerframework/framework/type/TypeFromTypeTreeVisitor.java
@@ -261,6 +261,9 @@ class TypeFromTypeTreeVisitor extends TypeFromTreeVisitor {
     if (elt instanceof TypeElement) {
       TypeElement typeElt = (TypeElement) elt;
       int idx = typeElt.getTypeParameters().indexOf(tpe);
+      if (idx == -1) {
+        idx = findIndex(typeElt.getTypeParameters(), tpe);
+      }
       ClassTree cls = (ClassTree) f.declarationFromElement(typeElt);
       if (cls == null || cls.getTypeParameters().isEmpty()) {
         // The type parameters in the source tree were already erased. The element already
@@ -299,6 +302,20 @@ class TypeFromTypeTreeVisitor extends TypeFromTreeVisitor {
     } else {
       throw new BugInCF("TypeFromTree.forTypeVariable: not a supported element: " + elt);
     }
+  }
+
+  private int findIndex(
+      List<? extends TypeParameterElement> typeParameters, TypeParameterElement type) {
+
+    TypeVariable typeVariable = (TypeVariable) type.asType();
+
+    for (int i = 0; i < typeParameters.size(); i++) {
+      TypeVariable typeVariable1 = (TypeVariable) typeParameters.get(i).asType();
+      if (TypesUtils.sames(typeVariable1, typeVariable)) {
+        return i;
+      }
+    }
+    return -1;
   }
 
   @Override

--- a/framework/src/main/java/org/checkerframework/framework/type/TypeFromTypeTreeVisitor.java
+++ b/framework/src/main/java/org/checkerframework/framework/type/TypeFromTypeTreeVisitor.java
@@ -311,7 +311,7 @@ class TypeFromTypeTreeVisitor extends TypeFromTreeVisitor {
 
     for (int i = 0; i < typeParameters.size(); i++) {
       TypeVariable typeVariable1 = (TypeVariable) typeParameters.get(i).asType();
-      if (TypesUtils.sames(typeVariable1, typeVariable)) {
+      if (TypesUtils.areSame(typeVariable1, typeVariable)) {
         return i;
       }
     }

--- a/framework/src/main/java/org/checkerframework/framework/type/TypeFromTypeTreeVisitor.java
+++ b/framework/src/main/java/org/checkerframework/framework/type/TypeFromTypeTreeVisitor.java
@@ -279,6 +279,9 @@ class TypeFromTypeTreeVisitor extends TypeFromTreeVisitor {
     } else if (elt instanceof ExecutableElement) {
       ExecutableElement exElt = (ExecutableElement) elt;
       int idx = exElt.getTypeParameters().indexOf(tpe);
+      if (idx == -1) {
+        idx = findIndex(exElt.getTypeParameters(), tpe);
+      }
       MethodTree meth = (MethodTree) f.declarationFromElement(exElt);
       if (meth == null || (idx == -1 && exElt.isVarArgs())) {
         // meth can be null when no source code was found for it.
@@ -304,6 +307,15 @@ class TypeFromTypeTreeVisitor extends TypeFromTreeVisitor {
     }
   }
 
+  /**
+   * Finds the index of {@code type} in {@code typeParameters} using {@link
+   * TypesUtils#areSame(TypeVariable, TypeVariable)} instead of {@link Object#equals(Object)}.
+   *
+   * @param typeParameters a list of type parameters
+   * @param type a type parameter
+   * @return the index of {@code type} in {@code typeParameters} using {@link
+   *     TypesUtils#areSame(TypeVariable, TypeVariable)} or -1 if it does not exist
+   */
   private int findIndex(
       List<? extends TypeParameterElement> typeParameters, TypeParameterElement type) {
 

--- a/framework/tests/all-systems/Issue6076.java
+++ b/framework/tests/all-systems/Issue6076.java
@@ -3,6 +3,7 @@ package issue6076;
 import java.util.Collection;
 import java.util.Set;
 
+
 public class Issue6076 {
 
   public SSTableReaderLoadingBuilder<BtiTableReader, BtiTableReader.Builder> loadingBuilder(
@@ -39,10 +40,12 @@ public class Issue6076 {
 
       public Builder(Descriptor descriptor) {}
 
+      @SuppressWarnings("unchecked")
       public B setTableMetadataRef(TableMetadataRef ref) {
         return (B) this;
       }
 
+      @SuppressWarnings("unchecked")
       public B setComponents(Collection<Component> components) {
         return (B) this;
       }

--- a/framework/tests/all-systems/Issue6076.java
+++ b/framework/tests/all-systems/Issue6076.java
@@ -3,7 +3,6 @@ package issue6076;
 import java.util.Collection;
 import java.util.Set;
 
-
 public class Issue6076 {
 
   public SSTableReaderLoadingBuilder<BtiTableReader, BtiTableReader.Builder> loadingBuilder(
@@ -34,6 +33,7 @@ public class Issue6076 {
         extends SSTableReader.Builder<R, B> {}
   }
 
+  @SuppressWarnings("all") // Just check for crashes.
   public abstract static class SSTable {
     public static class Builder<S extends SSTable, B extends Builder<S, B>> {
       public Builder() {}

--- a/framework/tests/all-systems/Issue6076.java
+++ b/framework/tests/all-systems/Issue6076.java
@@ -1,0 +1,63 @@
+package issue6076;
+
+import java.util.Collection;
+import java.util.Set;
+
+public class Issue6076 {
+
+  public SSTableReaderLoadingBuilder<BtiTableReader, BtiTableReader.Builder> loadingBuilder(
+      Descriptor descriptor, TableMetadataRef tableMetadataRef, Set<Component> components) {
+    return new BtiTableReaderLoadingBuilder(
+        new SSTable.Builder<>(descriptor)
+            .setTableMetadataRef(tableMetadataRef)
+            .setComponents(components));
+  }
+
+  public static class Component {}
+
+  public static class Descriptor {}
+
+  public static final class TableMetadataRef {}
+
+  public static class BtiTableReaderLoadingBuilder
+      extends SortedTableReaderLoadingBuilder<BtiTableReader, BtiTableReader.Builder> {
+    public BtiTableReaderLoadingBuilder(SSTable.Builder<?, ?> builder) {}
+  }
+
+  public static class BtiTableReader extends SSTableReaderWithFilter {
+    public static class Builder extends SSTableReaderWithFilter.Builder<BtiTableReader, Builder> {}
+  }
+
+  public abstract static class SSTableReaderWithFilter extends SSTableReader {
+    public abstract static class Builder<R extends SSTableReaderWithFilter, B extends Builder<R, B>>
+        extends SSTableReader.Builder<R, B> {}
+  }
+
+  public abstract static class SSTable {
+    public static class Builder<S extends SSTable, B extends Builder<S, B>> {
+      public Builder() {}
+
+      public Builder(Descriptor descriptor) {}
+
+      public B setTableMetadataRef(TableMetadataRef ref) {
+        return (B) this;
+      }
+
+      public B setComponents(Collection<Component> components) {
+        return (B) this;
+      }
+    }
+  }
+
+  public abstract static class SSTableReaderLoadingBuilder<
+      R extends SSTableReader, B extends SSTableReader.Builder<R, B>> {}
+
+  public abstract static class SortedTableReaderLoadingBuilder<
+          R extends SSTableReader, B extends SSTableReader.Builder<R, B>>
+      extends SSTableReaderLoadingBuilder<R, B> {}
+
+  public abstract static class SSTableReader extends SSTable {
+    public abstract static class Builder<R extends SSTableReader, B extends Builder<R, B>>
+        extends SSTable.Builder<R, B> {}
+  }
+}

--- a/javacutil/src/main/java/org/checkerframework/javacutil/TypesUtils.java
+++ b/javacutil/src/main/java/org/checkerframework/javacutil/TypesUtils.java
@@ -1273,6 +1273,10 @@ public final class TypesUtils {
    * name and the same enclosing element. Unlike {@link Types#isSameType(TypeMirror, TypeMirror)},
    * they do not have to be the same object.
    *
+   * <p>This method is needed when a type has some through type variable substitution, but only some
+   * of the type variables were substituted. Also, a new {@link TypeVariable} object is created as
+   * the type of a tree created by {@link org.checkerframework.javacutil.trees.TreeBuilder}.
+   *
    * @param typeVariable1 a type variable
    * @param typeVariable2 a type variable
    * @return if the two type variables are the same type variable

--- a/javacutil/src/main/java/org/checkerframework/javacutil/TypesUtils.java
+++ b/javacutil/src/main/java/org/checkerframework/javacutil/TypesUtils.java
@@ -34,6 +34,7 @@ import javax.lang.model.type.UnionType;
 import javax.lang.model.type.WildcardType;
 import javax.lang.model.util.Elements;
 import javax.lang.model.util.Types;
+import org.checkerframework.checker.interning.qual.EqualsMethod;
 import org.checkerframework.checker.nullness.qual.Nullable;
 import org.checkerframework.checker.signature.qual.BinaryName;
 import org.checkerframework.checker.signature.qual.CanonicalNameOrEmpty;
@@ -1265,5 +1266,17 @@ public final class TypesUtils {
     return unboxedKind == TypeKind.BYTE
         || unboxedKind == TypeKind.SHORT
         || unboxedKind == TypeKind.CHAR;
+  }
+
+  @EqualsMethod
+  public static boolean sames(TypeVariable key, TypeVariable other) {
+    if (key == other) {
+      return true;
+    }
+    Name otherName = other.asElement().getSimpleName();
+    Element otherEnclosingElement = other.asElement().getEnclosingElement();
+
+    return key.asElement().getSimpleName().contentEquals(otherName)
+        && otherEnclosingElement.equals(key.asElement().getEnclosingElement());
   }
 }

--- a/javacutil/src/main/java/org/checkerframework/javacutil/TypesUtils.java
+++ b/javacutil/src/main/java/org/checkerframework/javacutil/TypesUtils.java
@@ -1268,15 +1268,24 @@ public final class TypesUtils {
         || unboxedKind == TypeKind.CHAR;
   }
 
+  /**
+   * Returns true if the two type variables are the same type variable. Meaning they have the same
+   * name and the same enclosing element. Unlike {@link Types#isSameType(TypeMirror, TypeMirror)},
+   * they do not have to be the same object.
+   *
+   * @param typeVariable1 a type variable
+   * @param typeVariable2 a type variable
+   * @return if the two type variables are the same type variable
+   */
   @EqualsMethod
-  public static boolean sames(TypeVariable key, TypeVariable other) {
-    if (key == other) {
+  public static boolean areSame(TypeVariable typeVariable1, TypeVariable typeVariable2) {
+    if (typeVariable1 == typeVariable2) {
       return true;
     }
-    Name otherName = other.asElement().getSimpleName();
-    Element otherEnclosingElement = other.asElement().getEnclosingElement();
+    Name otherName = typeVariable2.asElement().getSimpleName();
+    Element otherEnclosingElement = typeVariable2.asElement().getEnclosingElement();
 
-    return key.asElement().getSimpleName().contentEquals(otherName)
-        && otherEnclosingElement.equals(key.asElement().getEnclosingElement());
+    return typeVariable1.asElement().getSimpleName().contentEquals(otherName)
+        && otherEnclosingElement.equals(typeVariable1.asElement().getEnclosingElement());
   }
 }

--- a/javacutil/src/main/java/org/checkerframework/javacutil/TypesUtils.java
+++ b/javacutil/src/main/java/org/checkerframework/javacutil/TypesUtils.java
@@ -1273,7 +1273,7 @@ public final class TypesUtils {
    * name and the same enclosing element. Unlike {@link Types#isSameType(TypeMirror, TypeMirror)},
    * they do not have to be the same object.
    *
-   * <p>This method is needed when a type has some through type variable substitution, but only some
+   * <p>This method is needed when a type has gone through type variable substitution, but only some
    * of the type variables were substituted. Also, a new {@link TypeVariable} object is created as
    * the type of a tree created by {@link org.checkerframework.javacutil.trees.TreeBuilder}.
    *

